- Tyler Arista, tja9@calvin.edu
- Student name, e-mail

# Instructions for today's practice

- Create a copy of this Jupyter Notebook and share it with your partner.
- Fill student names and e-mails in the text cell above.
- At the end of the practice, download the .ipynb file and upload it on Moodle.


**Thanks to prof. Ken Arnold for writing this practice.**

![](https://images.unsplash.com/photo-1502920514313-52581002a659?q=80&w=2067&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D)

# Dataset: Gapminder stats

[Gapminder](https://www.gapminder.org/data/) is a Swedish foundation whose goal is to "promote sustainable global development through the use of data and statistics". It is best known for its interactive visualizations that help illustrate global trends in areas such as health, income, population, and development.

We will be downloading data from their website and use it to practice reshaping and join operations with pandas.

Step by step instructions:

- Go to https://www.gapminder.org/data/
- Under “Choose individual indicators”, search for “GDP”. Click the one marked “Income”, it should say “GDP per capita (price and inflation adjusted, in PPP$2017)” when you click on it.
- At the top of the table, click Download As: CSV or XLSX. I got a file named lex.xlsx
- Click the header (“GDP per capita”) and select instead “Life Expectancy, at birth”. Download that as well. I got a file named gdp_pcap.xlsx

Now, import these xlsx files and load the data:

In [2]:
import pandas as pd

In [3]:
gapminder_gdp = pd.read_excel("gdp_pcap.xlsx")
gapminder_gdp.head()

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  ...  \
0  Afghanistan   599   599   599   599   599   599   599   599   599  ...   
1       Angola   465   466   469   471   472   475   477   479   481  ...   
2      Albania   585   587   588   590   592   593   595   597   598  ...   
3      Andorra  1710  1710  1710  1720  1720  1720  1730  1730  1730  ...   
4          UAE  1420  1430  1430  1440  1450  1450  1460  1460  1470  ...   

    2091   2092   2093   2094   2095   2096   2097   2098   2099   2100  
0   4800   4910   5030   5150   5270   5390   5520   5650   5780   5920  
1  24.8k  25.3k  25.9k  26.4k  26.9k  27.4k    28k  28.5k  29.1k  29.6k  
2    54k  54.6k  55.2k  55.8k  56.4k  56.9k  57.5k  58.1k  58.7k  59.2k  
3  79.3k  79.5k  79.8k  80.1k  80.4k  80.7k    81k  81.2k  81.5k  81.8k  
4  92.5k  92.6k  92.6k  92.7k  92.8k  92.9k  92.9k    93k  93.1k  93.1k  

[5 rows x 302 columns]

In [4]:
gapminder_life = pd.read_excel("lex.xlsx")
gapminder_life.head()

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  ...  \
0  Afghanistan  28.2  28.2  28.2  28.2  28.2  28.2  28.1  28.1  28.1  ...   
1       Angola  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  ...   
2      Albania  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  ...   
3      Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4          UAE  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  ...   

   2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  75.9  76.1  76.2  76.4  76.5  76.7  76.8  77.0  77.1  77.3  
1  75.8  75.9  76.0  76.1  76.3  76.4  76.5  76.7  76.8  76.9  
2  88.9  89.0  89.2  89.3  89.4  89.5  89.6  89.7  89.8  90.0  
3  91.0  91.1  91.3  91.4  91.5  91.6  91.7  91.8  91.9  92.0  
4  83.7  83.8  83.9  84.0  84.1  84.2  84.3  84.5  84.6  84.7  

[5 rows x 302 columns]

# From wide to long (melt)

Notice that the data is in “wide” format, with each year as a column. To join and plot this data, we’ll need it in “long” format, with each year as a row.

We'll use `pd.melt` to get it in this format. We'll need to tell it which columns form the primary key, specified by the `id_vars` argument. The other columns will be “melted” into a single column, specified by the `var_name` argument. The values for those columns will be put in a column specified by the `value_name` argument.

**📝 Exercise 1**: Fill in the blanks below with the appropriate column names. Then assign the result to a variable; I used `gdp_long`.

In [11]:
gdp_long = pd.melt(
    gapminder_gdp,
    id_vars=["country"],
    var_name="year", value_name="gdp_pcap"
)
gdp_long.head()

country  year gdp_pcap
0  Afghanistan  1800      599
1       Angola  1800      465
2      Albania  1800      585
3      Andorra  1800     1710
4          UAE  1800     1420

Now, repeat it for the life expectancy data (`life_long`).

In [12]:
life_long = pd.melt(
    gapminder_life,
    id_vars=['country'],
    var_name="year", value_name="Life Expectancy"
)
life_long.head()

country  year  Life Expectancy
0  Afghanistan  1800             28.2
1       Angola  1800             27.0
2      Albania  1800             35.4
3      Andorra  1800              NaN
4          UAE  1800             30.7

**📝 Exercise 2**: Answer: what are the **key columns** of the dataset both before it was melted (`gapminder_gdp`) and after (`gdp_long`)?

Before melting, the key columns in `gapminder_gdp` were "Country" and separate columns for each year (e.g., 1800, 1801), with GDP values in those year columns. After melting into `gdp_long`, the dataset has "Country," a single "Year" column, and a "GDP per Capita" column, where each row represents a country-year pair with its corresponding GDP value.


# Adjusting data types

Check the data types of the gdp_long table. There are two that are not what you’d expect.

In [13]:
gdp_long.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58695 entries, 0 to 58694
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   country   58695 non-null  object
 1   year      58695 non-null  object
 2   gdp_pcap  58695 non-null  object
dtypes: object(3)
memory usage: 1.3+ MB


Colum `year` is easy to fix:

In [14]:
gdp_long['year'] = gdp_long['year'].astype(int)
life_long['year'] = life_long['year'].astype(int)

A more tricky issue is that gdp_pcap uses k to mean “thousands”. For example:

In [15]:
gdp_long.tail()

country  year gdp_pcap
58690         Samoa  2100    29.2k
58691         Yemen  2100     8000
58692  South Africa  2100    50.2k
58693        Zambia  2100    19.6k
58694      Zimbabwe  2100    11.9k

We need to convert those k values to numbers. Here’s an approach:

In [16]:
def parse_number_with_units(num):
    if not isinstance(num, str):
        return num
    if num.endswith("k"):
        return float(num[:-1]) * 1000
    return float(num)

gdp_long['gdp_pcap'] = gdp_long['gdp_pcap'].map(parse_number_with_units)
gdp_long.tail()

country  year  gdp_pcap
58690         Samoa  2100   29200.0
58691         Yemen  2100    8000.0
58692  South Africa  2100   50200.0
58693        Zambia  2100   19600.0
58694      Zimbabwe  2100   11900.0

**📝 Exercise 3**: Search for the documentation about the `map()` function in `pandas`, which we used here, and try to answer: **what exactly does it do?** Describe it succintly. This function is very powerful and useful for wrangling! If you have more questions, we can talk about it later.

The map function in pandas is used to apply a function to each value in a column. So, it takes every value in the column, processes it through the function you provide, and returns the result for each value. This is really useful for changing/transforming data in a quick and easy way

# Joining

**📝 Exercise 4**: Now, join the datasets together. Call the result `gapminder`. This should have GDP and life expectancy of each country in each year.

Check for the total rows. You should have:
195 countries, 301 years = 58695 expected rows

In [19]:
gapminder = pd.merge(gdp_long, life_long, on=['country', 'year'])
print("Total # of rows:", len(gapminder))
gapminder.head()

Total # of rows: 58695


country  year  gdp_pcap  Life Expectancy
0  Afghanistan  1800     599.0             28.2
1       Angola  1800     465.0             27.0
2      Albania  1800     585.0             35.4
3      Andorra  1800    1710.0              NaN
4          UAE  1800    1420.0             30.7

# Adding regions to countries

Suppose we now want to visualize countries by region (Africa, Asia, etc). We need to perform another join, with another table!

The region data is available [here](https://www.gapminder.org/fw/four-regions/). `read_excel` can pull the data in directly for you. I'll save you some time by giving you that code:

In [20]:
region_data = pd.read_excel("https://docs.google.com/spreadsheets/d/1qHalit8sXC0R8oVXibc2wa2gY7bkwGzOybEMTWp-08o/export?format=xlsx", sheet_name="list-of-countries-etc")
region_data.head()

geo       name four_regions      eight_regions        six_regions  \
0  aus  Australia         asia  east_asia_pacific  east_asia_pacific   
1  brn     Brunei         asia  east_asia_pacific  east_asia_pacific   
2  khm   Cambodia         asia  east_asia_pacific  east_asia_pacific   
3  chn      China         asia  east_asia_pacific  east_asia_pacific   
4  fji       Fiji         asia  east_asia_pacific  east_asia_pacific   

  members_oecd_g77  Latitude  Longitude UN member since    World bank region  \
0             oecd     -25.0  135.00000      1945-11-01  East Asia & Pacific   
1              g77       4.5  114.66667      1984-09-21  East Asia & Pacific   
2              g77      13.0  105.00000      1955-12-14  East Asia & Pacific   
3              g77      35.0  105.00000      1945-10-24  East Asia & Pacific   
4              g77     -18.0  178.00000      1970-10-13  East Asia & Pacific   

  World bank, 4 income groups 2017 World bank, 3 income groups 2017  \
0                      High income                      High income   
1                      High income                      High income   
2              Lower middle income                    Middle income   
3              Upper middle income                    Middle income   
4              Upper middle income                    Middle income   

                  UNHCR  
0  Asia and the Pacific  
1  Asia and the Pacific  
2  Asia and the Pacific  
3  Asia and the Pacific  
4  Asia and the Pacific

**📝 Exercise 5**: We only want the column `four_regions`. So, rename it to `region`. Also, rename the `name` column so it has the same name as the corresponding column in the `gapminder` dataframe. To rename, use `rename(columns={"old_name": "new_name"})`.

You can also drop all the other columns with `region_data = region_data[["country", "region"]]`

In [21]:
region_data = region_data.rename(columns={"four_regions": "region", "name": "country"})
region_data = region_data[["country", "region"]]
region_data.head()

country region
0  Australia   asia
1     Brunei   asia
2   Cambodia   asia
3      China   asia
4       Fiji   asia

**📝 Exercise 6**: Now, join `region_data` with the `gapminder` data, and assign the final result to a variable (I used `gapminder_with_regions`).

In [23]:
gapminder_with_regions = pd.merge(gapminder, region_data, on="country")
print("Total rows after join:", len(gapminder_with_regions))
gapminder_with_regions.head()

Total rows after join: 57190


country  year  gdp_pcap  Life Expectancy region
0  Afghanistan  1800     599.0             28.2   asia
1  Afghanistan  1801     599.0             28.2   asia
2  Afghanistan  1802     599.0             28.2   asia
3  Afghanistan  1803     599.0             28.2   asia
4  Afghanistan  1804     599.0             28.2   asia

# Plot results



**📝 Exercise 7**: Let's now plot our data using plotly. Import the library and use a scatter plot with the following mappings:
- **x axis** = gdp per capita
- **y axis** = life expectancy
- **color** = region (Asia, America, etc)
- **animation_frame** = year (this will create an interactive slider)

Also, add the following arguments to your `px.scatter` function:
- `log_x=True` - this will set the x axis to logarithmic scale
- `range_x=[100,100000], range_y=[25,90]` - set the ranges of the x and y axes
- `hover_name="country"` - this will show the name of the country when you hover a data point
- `labels={"gdpPercap": "GDP per capita", "lifeExp": "Life expectancy (at birth)", "region": "region", "year": "Year"}` - set the labels
- `title="Life expectancy vs. GDP per capita, 1952-2007"` - set the title

In [26]:
import plotly.express as px

fig = px.scatter(
    gapminder_with_regions,
    x="gdp_pcap",
    y="Life Expectancy",
    color="region",
    animation_frame="year",
    hover_name="country",
    log_x=True,
    range_x=[100, 100000],
    range_y=[25, 90],
    labels={
        "GDP per Capita": "GDP per capita",
        "Life Expectancy": "Life expectancy (at birth)",
        "region": "Region",
        "Year": "Year"
    },
    title="Life expectancy vs. GDP per capita, 1952-2007"
)

fig.show()

**📝 Reflection Exercise**: Write a sentence or two of your overall
reflections on this practice. You may write whatever you want, but you
might perhaps respond to one or two of these questions:

-   Was anything unclear about this assignment?
-   How hard was it for you? Where did you get “stuck”?
-   How long did it take you?
-   What questions or uncertainties remain?
-   What skills do you think you’ll need more practice with?
-   Did you try anything out of curiosity that you weren’t specifically
    asked to do?

**Answer:** No, nothing was unclear about this assignment. I liked how the sections made it easier to follow and understand. This assignment was harder than the last two, but I enjoyed the challenge. Things like melting data and adjusting data types took a little more thought, but I was able to figure it out.